src/dns.c

```diff
@@ -390,6 +390,73 @@ int dns_encode_a_response(char *buf, size_t buflen, struct query *q)
 	return len;
 }
 
+int dns_encode_nxdomain(char *buf, size_t buflen, struct query *q, const char *zone)
+{
+    char rnamebuf[256];
+    char nsbuf[256];
+    HEADER *header;
+    char *p;
+
+    if (buflen < sizeof(HEADER))
+        return 0;
+
+    memset(buf, 0, buflen);
+    header = (HEADER*)buf;
+
+    header->id = htons(q->id);
+    header->qr = 1;         // response
+    header->opcode = 0;
+    header->aa = 1;         // authoritative
+    header->tc = 0;
+    header->rd = 0;
+    header->ra = 0;
+    header->rcode = 3;      // NXDOMAIN
+
+    header->qdcount = htons(1);
+    header->ancount = htons(0);
+    header->nscount = htons(1); // incluiremos SOA
+    header->arcount = htons(0);
+
+    p = buf + sizeof(HEADER);
+
+    // Question section
+    putname(&p, buflen - (p - buf), q->name);
+    CHECKLEN(4);
+    putshort(&p, q->type);
+    putshort(&p, C_IN);
+
+    // Authority section (SOA)
+    {
```

Comment:
is this extra scope needed?
The `startp` pointer could be declared earlier - also name it more descriptively (soa_start?). Declare variables before the code starts.

Reply:
No, I removed it and migrated the declaration to the top, renaming it to soa_start